### ISSUE TITLE
Why Fela?

### ISSUE THREAD

**Issue by koddsson** [tajo/cf-ui-next#1]
And not something like CSS modules?

**Comment by tajo** [comment]
**First of all, CSS Modules are great.** I used them a bit too and they solve almost all problems with CSS. But it's still CSS, so **you need to use some pre-processing / post-processing** and they are dependent on Webpack as well. CSS in JS is just JS, no additional build step needed.

**Some advantages of Fela over CSS Modules:**
- Sharing&replacing of global constants (things like font) is easy with themes.
- Better composition of styles via `Object.assign()` (+ I'm sure there another 100 different ways how to compose with JS).
- Dynamic CSS, all CSS rules are pure functions based on props. In CSS Modules, you would still have to toggle between different classes since all rules are static (not created in the runtime),
- Sharing JS vars with styles.
- Fela is using Atomic CSS, every single rule has its own class. In combination with predictive classnames (a, b, c .... a1, b2...) that should lead to smaller output styles. There are many optimizations possible on the background.
- Statically typed styles. I think that's going to be the next big thing in CSS world. Since Fela rulesets are just functions receiving props and returning objects we can specify their types. That should improve developer experience significantly (compiler will yell at you if you are trying to change a theme property that's going to be ignored by the component or if you do something silly as `paddingTop: 'red'` -> Catching errors automatically without needing a browser.

^^ I'm sure you could do most of these with CSS Modules + PostCSS as well but it probably wouldn't be that elegant. 

The real question is what the advantages of CSS Modules over Fela are (or CSS in JS in general).
- You could say it support "web platform" but that's not entirely true since you need Webpack and you really really want to pre/post process it even further.

Thoughts? @wyuenho @jwineman @hturan @terinjokes

**Comment by koddsson** [comment]
>The real question is what the advantages of CSS Modules over Fela are (or CSS in JS in general).

I think the CSS modules advantage over Fela would be that we don't have to re-write our CSS into JS structures (there might be some re-writes tho for things where we use the descendant selector).

>Sharing&replacing of global constants (things like font) is easy with themes.

If I'm understand this correctly, I think we can do themeing pretty straightforward in CSS modules by importing your theme in the container and passing it to the component.

>Dynamic CSS, all CSS rules are pure functions based on props. In CSS Modules, you would still have to toggle between different classes since all rules are static (not created in the runtime),

You can still pass down a `style` prop that would overwrite the specific rule in the CSS class.

I made a private repo with the css modules proposal and invited you and others to it, if you'd like to take a look :)

**Comment by hturan** [comment]
Hey @tajo, thanks so much for writing up your thoughts! CSS in JS is super exciting, and there are definite benefits for productivity, tooling and testing by having everything be dynamic.

Here are a few of my thoughts based on what you've written:

> CSS in JS is just JS, no additional build step needed.

One of our core requirements is a static external CSS file, so there will be changes to our build step when using either `fela` or `css-modules` (inclusion of `ExtractTextPlugin`, things like that). I definitely get the aversion to the Webpack requirement for `css-modules`, but it's already so central to everything we do on the front-end and will be required for `fela`, so I see its usage as a non-issue. 

> Sharing&replacing of global constants (things like font) is easy with themes

Theming in `css-modules` is pretty simple:
```
import styles from './Button.css';
const Button = ({type, theme = styles}) => (
  <button className={theme.button} type={type} />
)
```

> Better composition of styles using `Object.assign()`

[Composing styles](https://github.com/css-modules/css-modules#composition) is a core part of `css-modules`, although `{...styles, color}` could be cool in JS.

> Dynamic CSS, all CSS rules are pure functions based on props.

If this is the case, how will we extract static CSS at build time? Would this mean that when visiting the dashboard a user would get our bundled `styles.css`, our component with a `defaultTheme` and also our internal `cf-ui` theme override in JS? This seems *super* inefficient.

> Statically typed styles

Type-checked CSS would be amazing 🎉

---

There are a few benefits to using CSS Modules that have gone unmentioned:
- It's just CSS. Every member of the front-end team knows how to write CSS, and using a standard syntax helps understanding and on-boarding.
- Using our current PostCSS pipeline we get variables, calculation and constants.
- No performance impact
  - No repaint on `<style>` node updates
  - No processing `(props) => style` on every render
- Very little change to our build setup. Just import CSS, add `css-loader?modules` and `ExtractTextPlugin` to `webpack`, then we get a bundled `styles.css` output  ✨

My main concerns with CSS in JS is the extraction of a static CSS file. If we want dynamic styles as a function of props (`(props) => style`), then by it's very nature that's a runtime calculation, not a build-time one. There are a million solutions to this issue, but if we really want dynamic style calculation, there's nothing stopping us from using `css-modules` and then doing standard inline styles in JS(X):

`<button style={{backgroundColor: props.status === 'broken' ? 'red' : 'green'}} />`.

@tajo, could you update this repo with an example that spits out a static CSS file? Keep in mind that we don't have a `node` server currently, so just make `webpack` spit out the bundled JS and styles and `<link>` the CSS in 👍

@wyuenho, interested to get your take on this.

**Comment by wyuenho** [comment]
The reason fela uses some JS syntax to compose CSS is so it can control the generation of the class names, caches the values used and DCE. We can have a static "initial" CSS file injected into a style tag in the skeleton php file. This requires a weird deployment process tho.

1. Generate the initial CSS
2. have cf-www paste it into the skeleton
3. deploy cf-www
4. deploy www-next

`react-css-modules` maybe is more convenient, but in order to achieve what fela can do, it must have some CSS parsing capability.

Keep this in mind when you are playing with css-modules.

**Comment by wyuenho** [comment]
Also, I've said this before. I'd love to see some devtool profiling data for rapid state/style changes. It'd be a shame to speed up rendering with React and then regress back to hand-written code performance by introducing another tool.

**Comment by koddsson** [comment]
>DCE

What's DCE?

>Also, I've said this before. I'd love to see some devtool profiling data for rapid state/style changes. 

Could you elaborate on this?

**Comment by wyuenho** [comment]
DCE = Dead code elimination. You can take a look at fela's documentation. There are a few attempts. Styletron has a blog post about it as well.

For profiling, I want to see how much time the browser spends in JS and repaint if we say, change a button's state 1000 times.

**Comment by koddsson** [comment]
Are you wondering these things about fela or css-modules?

**Comment by wyuenho** [comment]
I'm saying you win some you lose some. Tho DCE is not required, it's nice
to have. I just want you think about these pros and cons when you are
comparing.

**Comment by hturan** [comment]
> ...in order to achieve what fela can do, it must have some CSS parsing capability

Right, in order to do dynamic things it needs to be written in a dynamic language. We're not comparing the *features* of `css-modules` and `fela` because that would be comparing CSS and JS. Instead, we're comparing how each solution would meet our needs as a business.

JS will always be more powerful than CSS, the question is whether we can take the performance hit and learning ramp. If we're happy to make those trade-offs, then full steam ahead.

(Also, keep in mind that this is a public space. Should we take this discussion internal?)

**Comment by koddsson** [comment]
>I'm saying you win some you lose some. Tho DCE is not required, it's nice to have. I just want you think about these pros and cons when you are comparing.

I mean, I am thinking about these things but I'm trying to find out if you have a specific concern about css-modules that I might be able to answer. I'm pretty sure that we can do DCE in css-modules pretty easily if it's not enabled by default as well as extracting into a seperate CSS file meaning that perf would be pretty good I'm guessing.

**Comment by wyuenho** [comment]
Meh. I'd take SASS any day over JS/CSS for its math expressions.

**Comment by tajo** [comment]
@koddsson I see you are overwriting things with `style` so that means you are using both CSS Modules and Reacts's CSS in JS which is not so good, because these are not extracted into `<style></style>` and also limited in functionality (no pseudo classes etc).

It seems to me that there is a big gap between default cf-ui styles and our internal styles, so we will be overwriting a lot of rules all the time.

@hturan I'm going to update this demo later this week and try to answer more things.

**Comment by koddsson** [comment]
>I see you are overwriting things with style so that means you are using both CSS Modules and Reacts's CSS in JS which is not so good, because these are not extracted into <style></style> and also limited in functionality (no pseudo classes etc).

Yeah just for the dynamic stuff, right? So it wouldn't be extracted in any way? If we care about extracting dynamic styles into a file we could add a BEM class like `Button--red` or whatever and pass it down to buttons dynamically.

**Comment by tajo** [comment]
Ok, dynamic stuff is a whole different thing.

**My point**
1. Button has a complete set of default styles and these styles are inside of the Button component.
2. We want to customize it but we just need to change 1/3 of default styles because the rest can be reused. So in my case, we can wrap the Button with ThemeProvider and pass an object with styles that we want to change.

How do you overwrite default styles with our custom styles? Where will we keep .css files? **Do we need identical .css files for default theme and cf theme (so there will be duplicate rules)?**

**Comment by koddsson** [comment]
We could achieve it like this:

```css
/* Button.css */
.Button {
  background-color: gray;
  border: 1px solid black;
  margin: 10px;
}
```

```js
/* Button.js */
import React from 'react'
import styles from './Button.css' // These are the external styles

export class Button extends React.Component {
  render () {
    const {className, ...props} = this.props;
    const classes = [styles.Button];
    if (className) {
      classes.push(className);
    }
    return (
      <button className={classes.join(' ')} {...props}>
        {this.props.children}
      </button>
    )
  }
}
```

```css
/* super-secret-css/Button.css */

.Button {
  background-color: orange;
}
```

```js
/* SomeContainer.js */
import React from 'react'
import {Button} from 'cf-component-button';

import styles from './super-secret-css/Button.css' // These are the internal styles

export class Button extends React.Component {
  render () {
    return (
      <div>
        <Button className={styles.Button}>
          Click here to orange yourself
        </Button>
      </div>
    );
  }
}
```

But that only applies if reusing the external styles is a thing we really wanna do. We currently aren't doing it.

**Comment by koddsson** [comment]
(I did not test the above code at all, so it could contain errors, factual or syntax)

**Comment by tajo** [comment]
Cool. Now it's more realistic. What if you need to combine three different themes (twitter bootstrap + cf-ui default + cf-ui private)?

I know you can do it but `Object.assign()` is a bit nicer. :-)
Also, your solution would produce a bigger stylesheet because overwritten rules are not eliminated.
